numpy/core/tests/test_nditer.py

```diff
@@ -1961,10 +1963,45 @@ def test_iter_buffered_cast_structured_type():
     sdt2 = [('b', 'O'), ('a', 'f8')]
     a = np.array([(1, 2, 3), (4, 5, 6)], dtype=sdt1)
 
-    assert_raises(ValueError, lambda : (
-        nditer(a, ['buffered', 'refs_ok'], ['readwrite'],
-               casting='unsafe',
-               op_dtypes=sdt2)))
+    for intent in ["readwrite", "readonly", "writeonly"]:
+        # If the following assert fails, the place where the error is raised
+        # within nditer may change. That is fine, but it may make sense for
+        # a new (hard to design) test to replace it:
+        assert np.can_cast(a.dtype, sdt2, casting="unsafe")
+        simple_arr = np.array([1, 2], dtype="i")  # succeeds but needs clean up
+        with pytest.raises(ValueError):
+            nditer((simple_arr, a), ['buffered', 'refs_ok'], [intent, intent],
+                   casting='unsafe', op_dtypes=["f", sdt2])
```

Comment:
I was a bit surprised, but this should cover the error cleanup path codecov complained about (we will see).  This is almost impossible to test, because it is not well defined that the error will occur at that point, to be sure, we would basically have to create a maliciously designed `dtype` from scratch.